hazelcast/src/hazelcast/client/spi.cpp

```diff
@@ -1325,22 +1342,29 @@ namespace hazelcast {
                 }
 
                 std::shared_ptr<connection::Connection> ClientInvocation::get_send_connection() const {
-                    return *send_connection_.load();
+                    return send_connection_.load()->lock();
                 }
 
                 std::shared_ptr<connection::Connection> ClientInvocation::get_send_connection_or_wait() const {
-                    while (!(*send_connection_.load()) && lifecycle_service_.is_running()) {
-                        std::this_thread::yield();
-                    }
+                    std::shared_ptr<connection::Connection> connection;
+                    do {
+                        auto loaded_conn = send_connection_.load();
+                        if (loaded_conn) {
+                            connection = loaded_conn->lock();
+                        }
+                        if (!connection) {
+                            std::this_thread::yield();
+                        }
+                    } while (lifecycle_service_.is_running() && !connection);
```

Comment:
Fixed it. changed the method name to `wait_invoked` and method signature to not return anything.

Reply:
No, it still not doing the same thing as before. See the constructor of ClientInvocation. It calls set_send_connection(nullptr) and that stores a **nonempty** shared_ptr that points to an empty weak_ptr. The condition `!send_connection_.load()` in the while loop will always evaluate to false and the loop will never run. Maybe you can modify the ctor so that the send_connection_ is initially empty.

Reply:
Should be fixed now. Can you please double check?